src/main/scala/loamstream/util/BiMap.scala

```diff
@@ -0,0 +1,79 @@
+package loamstream.util
+
+/**
+ * @author clint
+ * Nov 7, 2018
+ * 
+ * Represents a bi-directional one-to-one mapping between As and Bs, with unique keys and values.
+ * Allows efficient lookups (constant-time, assuming the component maps offer constant-time lookups) by 
+ * keys /and/ values.
+ * 
+ * Note: uniqueness of keys and values is guaranteed by overwriting existing mappings, just like with regular 
+ * Maps.  When adding mappings with `+`, this class ensures that both forward and backward mappings
+ * are overwritten if they're present.
+ */
+final case class BiMap[A, B] private (private[util] val forward: Map[A, B], private[util] val backward: Map[B, A]) {
```

Comment:
Just asking to make 100% sure... unless this is some "special edition" BiMap, the reason bi-directional maps are hard to use is that the data is so rarely a guaranteed 1:1 mapping as opposed to 1:n (the same value may appear multiple times in the map). I just want to make sure this is guaranteed 1:1 and you've taken that into consideration? And that any unit-testing you have in place for it tests against it to make sure it can't happen and have a key overridden by accident that will break the graph?

Reply:
Yes, there's a test specifically for that case, and the guard in the constructor provides a basic runtime soundness check.

Reply:
Hmm... I don't know how big they get, but I'd be concerned with this turning into a performance problem down the road as opposed to a check on insert? 

My thinking is this: 

On insert, it's O(1) to check if the value is already present for a different key. That test is simple. However, in the constructor you're checking *everything* (I'm assuming) to make sure it's OK. That's O(n) for every instance created.

If the `BiMap` is only ever created once and then used to do stuff, then cool; the constructor is the easier place to do that work (and makes the tests easier).

If, however, once created, you're adding to it after-the-fact and creating new `BiMap` instances, then the constructor is likely not a great place.

Reply:
> On insert, it's O(1) to check if the value is already present for a different key. That test is simple. However, in the constructor you're checking everything (I'm assuming) to make sure it's OK. That's O(n) for every instance created.

That's something I specifically avoided for that very reason.  The runtime guard is pretty basic - it just checks that the forward and backward maps are the same size, since any difference means the map is unsound, and getting sizes is cheap.  

There are likely (maybe?) ways for a `BiMap` to be unsound with equally-sized forward and backward maps, and for that, the code that adds and removes mappings jumps through some hoops to make sure that duplicate values are removed from both the forward and backward maps in the case any mappings need to be overwritten, and this is unit tested, both for duplicate keys (which was there before) and for duplicate values (which you caught).